### PR TITLE
planner: fix compatibility bug for window function

### LINKED ISSUE #11010: window function: Window with RANGE N PRECEDING/FOLLOWING frame requires exactly one ORDER BY expression, of numeric or temporal type
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```mysql
SELECT DENSE_RANK() OVER w1 AS 'dense_rank', fieldA, fieldB FROM ( SELECT `col_time` AS fieldA, `col_time` AS fieldB FROM `table70_int_autoinc` ) as t WINDOW w1 AS (PARTITION BY fieldB ORDER BY fieldB DESC, fieldA DESC RANGE BETWEEN CURRENT ROW AND 1250951168 FOLLOWING);
```

2. What did you expect to see?

```
ERROR 3587 (HY000): Window 'w1' with RANGE N PRECEDING/FOLLOWING frame requires exactly one ORDER BY expression, of numeric or temporal type
```

3. What did you see instead?

```mysql
+------------+----------+----------+
| dense_rank | fieldA   | fieldB   |
+------------+----------+----------+
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | NULL     | NULL     |
|          1 | 00:20:00 | 00:20:00 |
|          1 | 00:20:00 | 00:20:00 |
|          1 | 00:20:02 | 00:20:02 |
|          1 | 00:20:02 | 00:20:02 |
|          1 | 00:20:03 | 00:20:03 |
|          1 | 00:20:03 | 00:20:03 |
|          1 | 00:20:04 | 00:20:04 |
|          1 | 00:20:04 | 00:20:04 |
|          1 | 00:20:04 | 00:20:04 |
|          1 | 00:20:04 | 00:20:04 |
|          1 | 00:20:04 | 00:20:04 |
|          1 | 00:20:05 | 00:20:05 |
|          1 | 00:20:05 | 00:20:05 |
|          1 | 00:20:06 | 00:20:06 |
|          1 | 00:20:06 | 00:20:06 |
|          1 | 00:20:07 | 00:20:07 |
|          1 | 00:20:07 | 00:20:07 |
|          1 | 00:20:07 | 00:20:07 |
|          1 | 00:20:08 | 00:20:08 |
|          1 | 00:20:08 | 00:20:08 |
|          1 | 00:20:09 | 00:20:09 |
|          1 | 00:20:09 | 00:20:09 |
|          1 | 00:33:22 | 00:33:22 |
|          1 | 01:52:29 | 01:52:29 |
|          1 | 01:57:37 | 01:57:37 |
|          1 | 03:20:03 | 03:20:03 |
|          1 | 03:37:07 | 03:37:07 |
|          1 | 04:31:41 | 04:31:41 |
|          1 | 04:54:55 | 04:54:55 |
|          1 | 05:00:38 | 05:00:38 |
|          1 | 05:13:27 | 05:13:27 |
|          1 | 05:55:46 | 05:55:46 |
|          1 | 07:58:07 | 07:58:07 |
|          1 | 08:30:20 | 08:30:20 |
|          1 | 09:18:48 | 09:18:48 |
|          1 | 10:24:53 | 10:24:53 |
|          1 | 10:55:41 | 10:55:41 |
|          1 | 11:12:03 | 11:12:03 |
|          1 | 11:18:21 | 11:18:21 |
|          1 | 11:44:29 | 11:44:29 |
|          1 | 11:52:45 | 11:52:45 |
|          1 | 11:57:50 | 11:57:50 |
|          1 | 12:47:08 | 12:47:08 |
|          1 | 13:07:05 | 13:07:05 |
|          1 | 15:26:51 | 15:26:51 |
|          1 | 15:32:13 | 15:32:13 |
|          1 | 15:56:16 | 15:56:16 |
|          1 | 16:10:57 | 16:10:57 |
|          1 | 16:16:42 | 16:16:42 |
|          1 | 16:22:06 | 16:22:06 |
|          1 | 16:30:12 | 16:30:12 |
|          1 | 19:57:25 | 19:57:25 |
|          1 | 20:32:24 | 20:32:24 |
|          1 | 22:03:23 | 22:03:23 |
|          1 | 22:12:18 | 22:12:18 |
|          1 | 23:02:56 | 23:02:56 |
|          1 | 23:21:44 | 23:21:44 |
|          1 | 23:38:42 | 23:38:42 |
|          1 | 23:59:09 | 23:59:09 |
+------------+----------+----------+
70 rows in set, 1 warning (0.01 sec)

mysql> show warnings;
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                                      |
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Note  | 3599 | Window function 'DENSE_RANK' ignores the frame clause of window 'w1' and aggregates over the whole partition |
+-------+------+--------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
Fix #11010, #11001, #11002, #11011.

### What is changed and how it works?
The error judging should not be involved in the build function. It caused the wrong frame like the default frame to check. That's a fault.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - None

Related changes

 - Need to cherry-pick to the release branch
 - Need to change common test

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3128,6 +3128,11 @@ func (b *PlanBuilder) buildWindowFunctionFrameBound(ctx context.Context, spec *a
 	col := orderByItems[0].Col
 	isNumeric, isTemporal := types.IsTypeNumeric(col.RetType.Tp), types.IsTypeTemporal(col.RetType.Tp)
 	if !isNumeric && !isTemporal {
+		// For compatibility, we should ingore error for DENSE_RANK in this case.
+		if spec.Name.O == "DENSE_RANK" {
```
Comment (lamxTyler) [inline planner/core/logical_plan_builder.go]: Even without this branch, the warnings will also be raised.

**Comment by winoros** [comment]
I think the codes are not special for `DENSE_RANK`. It's for all window functions right?

**Comment by lzmhhh123** [comment]
@winoros Yes. It is better to change the description.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -3380,17 +3381,27 @@ func (b *PlanBuilder) handleDefaultFrame(spec *ast.WindowSpec, windowFuncName st
...
 	if !needFrame && spec.Frame != nil {
+		if spec.OrderBy != nil && len(spec.OrderBy.Items) > 1 && spec.Frame.Type == ast.Ranges {
+			start, end := spec.Frame.Extent.Start, spec.Frame.Extent.End
+			if start.Type == ast.Preceding && !start.UnBounded {
+				return spec, false, ErrWindowRangeFrameOrderType.GenWithStackByArgs(getWindowName(spec.Name.O))
+			}
+			if end.Type == ast.Following && !end.UnBounded {
+				return spec, false, ErrWindowRangeFrameOrderType.GenWithStackByArgs(getWindowName(spec.Name.O))
+			}
+		}
```
Comment (SunRunAway) [inline planner/core/logical_plan_builder.go]: Your code is for frame validation, is that suitable to be put in `handleDefaultFrame`?

Comment (SunRunAway) [reply planner/core/logical_plan_builder.go]: Besides, the ErrWindowRangeFrameOrderType error is `needFrame` irrelevant, considering this case
```mysql
SELECT FIRST_VALUE(fieldA) OVER w1 AS 'dense_rank', fieldA, fieldB  FROM ( SELECT `col_time` AS fieldA, `col_time` AS fieldB FROM `table20_int_autoinc` ) as t     WINDOW w1 AS ( PARTITION BY fieldB  ORDER BY fieldB DESC, fieldA DESC  RANGE BETWEEN CURRENT ROW AND 1 FOLLOWING);
```

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: Actually, the `handleDefaultFrame` is a function to remove the invalidated frames. So it's reasonable to put it here.

Comment (lamxTyler) [reply planner/core/logical_plan_builder.go]: And there are much more frame validations in `buildWindowFunctionFrame`.

Comment (SunRunAway) [reply planner/core/logical_plan_builder.go]: The reasonable order for building a window function is
1) build from ast
2) validate
3) handleDefaultFrame

May also fix #11001 #11002, and make #11011 easy to fix.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: Now, the frame validations are after `groupWindowFuncs`. We have to `handleDefaultFrame` in `groupWindowFuncs`. It hard to move this validation to `buildWindowFunctionFrame`...

**Pull request merged**